File: shared/btagent_shared/types/pattern_hunt.py
"""Cross-Investigation Pattern Hunter schemas (Phase 6 #120).

The cross-investigation counterpart to the per-entity Behavioral Hunter
(#114) and the rule-driven Hunt Pack Runner (#112). Where those hunt a
single live telemetry stream, this hunts the *corpus of closed
investigations*: the same observable (an IOC, a TLD, a command-line
fragment, an asset, an ASN, a technique) showing up faintly across many
**unrelated** past cases is a weak signal that a slow campaign has been
tunnelling under the SOC's radar.

A :class:`WeakSignal` is one such observable with its source-investigation
provenance; a :class:`WeakSignalCluster` groups similar signals and carries
the explainable rank; a :class:`PatternHuntProposal` turns a high-ranking
cluster into a ready-to-run :class:`~btagent_shared.types.hunt.HuntInput`.

Design notes (kept in lock-step with the rest of ``btagent_shared.types``):

1. **Zero heavy dependencies** — Pydantic only, so the engine / backend /
   tests can all import this tier without pulling LangChain, MCP, etc.
2. **``extra="forbid"`` everywhere** — a typo'd field is a contract bug, not
   a silently-dropped value.
3. **Enum values are lowercase strings** to match ``Severity`` / ``IOCType``
   / ``HuntDomain`` conventions.
4. The dependency-free extraction + clustering + proposal logic lives in
   :mod:`btagent_shared.hunt.pattern`; persistence + corpus-walking live in
   ``backend/btagent_backend/services/pattern_hunt_service.py``.
"""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field

from btagent_shared.types.hunt import HuntInput


class WeakSignalKind(StrEnum):
    """The class of observable a weak signal is keyed on.

    Stored as a lowercase string so adding a kind is a non-destructive
    change. Each kind maps to a distinct extraction path in
    :class:`btagent_shared.hunt.pattern.WeakSignalExtractor`.
    """

    IOC = "ioc"
    TLD = "tld"
    CMDLINE_FRAGMENT = "cmdline_fragment"
    ASSET = "asset"
    ASN = "asn"
    TECHNIQUE = "technique"


class ProposalState(StrEnum):
    """Lifecycle of a :class:`PatternHuntProposal`.

    * ``PROPOSED`` — surfaced by a scan, awaiting analyst review.
    * ``ACCEPTED`` — analyst kicked off the hunt from this proposal.
    * ``DISMISSED`` — analyst marked it not-interesting; the service
      down-weights similar clusters in future scans so the same shape
      doesn't keep re-surfacing.
    * ``SNOOZED`` — temporarily hidden; also down-weights similar future
      surfacing but is explicitly reversible.
    """

    PROPOSED = "proposed"
    ACCEPTED = "accepted"
    DISMISSED = "dismissed"
    SNOOZED = "snoozed"


class ProposalOutcome(StrEnum):
    """Terminal result of a hunt launched from a proposal (Phase B feedback).

    Persisted here so the closed-loop tuning that lands in Phase B has a
    column to write into; Phase A only ever sets it to ``None``.
    """

    CLEAN = "clean"
    HIT = "hit"


# --------------------------------------------------------------------------- #
# Core domain models
# --------------------------------------------------------------------------- #


class WeakSignal(BaseModel):
    """A single faint observable extracted from the closed-investigation corpus.

    The ``distinct_investigation_count`` is deliberately a stored field
    (rather than ``len(investigation_refs)``): cross-investigation diversity
    is the load-bearing ranking term (see
    :class:`btagent_shared.hunt.pattern.WeakSignalClusterer`), so it is
    computed once at extraction time and pinned, never re-derived from a
    possibly-truncated ref list.
    """

    model_config = ConfigDict(extra="forbid")

    id: str
    kind: WeakSignalKind
    value: str = Field(..., min_length=1, max_length=2048)
    # The original :class:`~btagent_shared.types.enums.IOCType` value (e.g.
    # ``"ip"`` / ``"url"`` / ``"hash_sha256"``) for ``kind == IOC`` signals, so
    # the downstream proposal transformer can preserve the known indicator type
    # instead of flattening every exact IOC to ``OTHER`` (which would make the
    # hypothesis generator emit zero hypotheses). ``None`` for non-IOC kinds and
    # for genuinely unknown IOC types (which fall back to ``OTHER`` downstream).
    ioc_type: str | None = Field(
        default=None,
        max_length=30,
        description="Original IOCType value for IOC-kind signals; None otherwise.",
    )
    first_seen: datetime
    last_seen: datetime
    investigation_refs: list[str] = Field(
        default_factory=list,
        description="Investigation ids this signal was observed in (de-duplicated).",
    )
    distinct_investigation_count: int = Field(
        default=0,
        ge=0,
        description="Number of distinct investigations — the diversity term in ranking.",
    )


class WeakSignalCluster(BaseModel):
    """A group of similar weak signals plus its explainable rank.

    ``score`` is produced by the pure clusterer as
    ``frequency × recency × cross-investigation diversity``; ``rationale``
    is the human-readable breakdown of those factors so an analyst can see
    *why* a cluster rose to the top without reading the source code.
    """

    model_config = ConfigDict(extra="forbid")

    id: str
    members: list[WeakSignal] = Field(default_factory=list)
    score: float = Field(default=0.0, ge=0.0)
    rationale: str = Field(default="", max_length=4096)


class PatternHuntProposal(BaseModel):
    """A high-ranking cluster turned into a ready-to-run hunt.

    Mirrors :class:`btagent_backend.db.models_pattern.PatternHuntProposalRow`.
    The :class:`~btagent_shared.types.hunt.HuntInput` is fully-formed and
    guaranteed non-empty in at least one of (ttps, iocs, adversaries) by the
    proposal transformer.
    """

    model_config = ConfigDict(extra="forbid")

    id: str
    org_id: str
    cluster_id: str
    hunt_input: HuntInput
    rationale: str = Field(default="", max_length=4096)
    state: ProposalState = ProposalState.PROPOSED
    outcome: ProposalOutcome | None = None
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/pattern_hunt.ts
/**
 * Pattern Hunt UI types (#120 Phase B).
 *
 * Mirrors the backend shapes in
 * ``shared/btagent_shared/types/pattern_hunt.py`` — kept in sync manually;
 * the string literal unions match the backend StrEnum values exactly.
 */

// --------------------------------------------------------------------------- //
// Enums (mirrored from Python StrEnum)
// --------------------------------------------------------------------------- //

export type WeakSignalKind =
  | "ioc"
  | "tld"
  | "cmdline_fragment"
  | "asset"
  | "asn"
  | "technique";

export type ProposalState = "proposed" | "accepted" | "dismissed" | "snoozed";

export type ProposalOutcome = "clean" | "hit";

// --------------------------------------------------------------------------- //
// Domain models
// --------------------------------------------------------------------------- //

/**
 * A single faint observable extracted from the closed-investigation corpus.
 * The UI uses this to show the source signal members inside a proposal card.
 */
export interface WeakSignal {
  id: string;
  kind: WeakSignalKind;
  value: string;
  ioc_type: string | null;
  first_seen: string;
  last_seen: string;
  investigation_refs: string[];
  distinct_investigation_count: number;
}

/**
 * A group of similar weak signals with its explainable rank.
 *
 * ``score`` is ``frequency × recency × cross-investigation diversity``;
 * ``rationale`` is the human-readable breakdown the UI surfaces in the
 * "why did this surface?" section.
 */
export interface WeakSignalCluster {
  id: string;
  members: WeakSignal[];
  score: number;
  rationale: string;
}

/**
 * A high-ranking cluster turned into a ready-to-run hunt proposal.
 *
 * ``hunt_input`` is the serialised ``HuntInput`` (adversaries / ttps / iocs /
 * scope) that Phase C will use to emit a ``HuntPlan``; Phase B just shows it
 * for context.
 */
export interface PatternHuntProposal {
  id: string;
  org_id: string;
  cluster_id: string;
  score: number;
  hunt_input: {
    adversaries: string[];
    ttps: string[];
    iocs: Array<{
      type: string;
      value: string;
      [key: string]: unknown;
    }>;
    scope?: {
      environments: string[];
      hosts: string[];
      date_from: string | null;
      date_to: string | null;
      backends: string[];
    };
  };
  rationale: string;
  /** Analyst triage notes accumulated across transitions (kept separate from
   * the generated `rationale` so that text stays pristine). */
  triage_rationale: string;
  state: ProposalState;
  outcome: ProposalOutcome | null;
  created_at: string;
  updated_at: string;
}

// --------------------------------------------------------------------------- //
// Request / response payloads
// --------------------------------------------------------------------------- //

export interface ActionRequest {
  rationale?: string;
}

export interface PatternHuntProposalListResponse {
  items: PatternHuntProposal[];
  total: number;
}

// --------------------------------------------------------------------------- //
// UI-only aggregated / derived types
// --------------------------------------------------------------------------- //

/**
 * A display-oriented summary of the hunt-input signal types for a proposal.
 * Derived client-side from ``hunt_input`` so the card can render chip groups.
 */
export interface ProposalSignalSummary {
  ttp_count: number;
  ioc_count: number;
  adversary_count: number;
}

/** Filter tab union — "all" is the catch-all; others match ProposalState. */
export type ProposalFilter = ProposalState | "all";

// --------------------------------------------------------------------------- //
// HuntPlan (#120 Phase C — compiled plan + execution)
// --------------------------------------------------------------------------- //

/** Compile lifecycle of the stored plan row (NOT the plan's execution state). */
export type HuntPlanRowStatus = "pending" | "ready" | "failed";

/**
 * ``GET /pattern/proposals/{id}/plan`` response — compile status plus the
 * serialised HuntPlan. ``plan`` is null until the compile finishes; after an
 * execution it additionally carries a ``last_run`` summary object.
 */
export interface ProposalHuntPlan {
  id: string;
  org_id: string;
  proposal_id: string;
  status: HuntPlanRowStatus;
  plan: {
    id: string;
    state: string;
    hypotheses: Array<{ ttp_id: string; ttp_name: string; priority: number }>;
    ttp_entries: Array<{
      ttp_id: string;
      ttp_name: string;
      queries: Record<string, { backend: string; query: string }>;
    }>;
    last_run?: {
      run_id: string;
      started_at: string;
      completed_at: string | null;
      findings_created: number;
      error_count: number;
      per_ttp: Record<string, { hits: number; errors: string[] }>;
    };
    [key: string]: unknown;
  } | null;
  error: string;
  created_at: string;
  updated_at: string;
}

/** ``POST /pattern/proposals/{id}/plan/execute`` response. */
export interface ExecutePlanResponse {
  plan: ProposalHuntPlan;
  queued: boolean;
  findings_created: number | null;
}
